play-services-nearby-core/src/main/kotlin/org/microg/gms/nearby/exposurenotification/ExposureNotificationServiceImpl.kt

```diff
@@ -25,18 +26,53 @@ import org.json.JSONObject
 import org.microg.gms.common.Constants
 import org.microg.gms.common.PackageUtils
 import org.microg.gms.nearby.exposurenotification.Constants.*
+import org.microg.gms.nearby.exposurenotification.proto.TEKSignatureList
 import org.microg.gms.nearby.exposurenotification.proto.TemporaryExposureKeyExport
 import org.microg.gms.nearby.exposurenotification.proto.TemporaryExposureKeyProto
 import java.io.File
 import java.io.InputStream
+import java.security.KeyFactory
 import java.security.MessageDigest
+import java.security.Signature
+import java.security.spec.X509EncodedKeySpec
+import java.util.zip.ZipEntry
 import java.util.zip.ZipFile
 import kotlin.math.max
 import kotlin.math.roundToInt
 import kotlin.random.Random
 
 class ExposureNotificationServiceImpl(private val context: Context, private val lifecycle: Lifecycle, private val packageName: String) : INearbyExposureNotificationService.Stub(), LifecycleOwner {
 
+    // Table of back-end public keys, used to verify the signature of the diagnosed TEKs.
+    // The table is indexed by package names.
+    private val backendPubKeyForPackage = mapOf<String, String>(
+            Pair("ch.admin.bag.dp3t.dev",
+                    "MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEsFcEnOPY4AOAKkpv9HSdW2BrhUCWwL15Hpqu5zHaWy1Wno2KR8G6dYJ8QO0uZu1M6j8z6NGXFVZcpw7tYeXAqQ=="),
+            Pair("ch.admin.bag.dp3t.test",
+                    "MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEsFcEnOPY4AOAKkpv9HSdW2BrhUCWwL15Hpqu5zHaWy1Wno2KR8G6dYJ8QO0uZu1M6j8z6NGXFVZcpw7tYeXAqQ=="),
+            Pair("ch.admin.bag.dp3t.abnahme",
+                    "MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEsFcEnOPY4AOAKkpv9HSdW2BrhUCWwL15Hpqu5zHaWy1Wno2KR8G6dYJ8QO0uZu1M6j8z6NGXFVZcpw7tYeXAqQ=="),
+            Pair("ch.admin.bag.dp3t",
+                    "MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEK2k9nZ8guo7JP2ELPQXnUkqDyjjJmYmpt9Zy0HPsiGXCdI3SFmLr204KNzkuITppNV5P7+bXRxiiY04NMrEITg=="),
+        )
+
+    // Back-end public key for this package
+    private val backendPublicKey = backendPubKeyForPackage[packageName]?.let {
+        try {
+            KeyFactory.getInstance("EC").generatePublic(X509EncodedKeySpec(Base64.decode(it, Base64.DEFAULT)))
+        } catch (e: Exception) {
+            Log.w(TAG, "Failed to retrieve back-end public key for ${packageName}: " + e.message)
+            null
+        }
+    }
+
+    // Table of supported signature algorithms for the diagnosed TEKs.
+    // The table is indexed by ASN.1 OIDs as specified in https://tools.ietf.org/html/rfc5758#section-3.2
+    private val sigAlgoForOid = mapOf<String, Signature>(
+            Pair("1.2.840.10045.4.3.2", Signature.getInstance("SHA256withECDSA")),
```

Comment:
Either map these to the name string (e.g. "SHA256withECDSA") and only run `Signature.getInstance()` when needed or map to a function that creates instances of Signature (by surrounding the invocation of `Signature.getInstance()` with curly braces).

PS: I very much like the notation of kotlin infix to for map creation, like
```kotlin
val sigAlgoForOid = mapOf<String, Function0<Signature>>(
  "1.2.840.10045.4.3.2" to { Signature.getInstance("SHA256withECDSA") },
  "1.2.840.10045.4.3.4" to { Signature.getInstance("SHA512withECDSA") }
)
```

Reply:
I changed it as you suggested (indeed, the `... to ...` notation is quite nice, changed my other map accordingly :) ).

It is a trade-off though, exchanging the unused instances in memory for a new instantiation for every signature verification, right?